### ISSUE TITLE
7.1-Tips for Training: Batch Normalization

### ISSUE THREAD

**Issue by Pin-Jiun** [Pin-Jiun/Machine-Learing-HYL#45]
當error surface表現的很崎嶇, 會比較難train, 這時候就可以使用Normalization的方式把山給剷平

Optimization的問題一直是很重要的, 儘管是類似一個碗狀的surface也會很難train
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/1612952f-0e2f-471f-a40f-7e05db222b5b)

在 W1 這個方向上面 你的斜率變化很小 在 W2 這個方向上面斜率變化很大
你今天如果是固定的 learning rate 你可能很難得到好的結果

到底是從什麼地方來的?
那我們這邊就是舉一個例子

假設我現在有一個非常非常非常簡單的 model
它的輸入是 x1 跟 x2 x1 跟 x2 它對應的參數就是 W1 跟 W2
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/0fcd9ac7-a2c3-4e32-bf84-723620f2c080)

它是一個 linear 的 model 沒有 activation function W1 乘 x1 W2 乘 x2 加上 b 以後就得到 y
然後會計算 y 跟 y hat 之間的差距當做 e 把所有 training data e 加起來 = L

當我們對 W1 有一個小小的改變 比如說加上 delta W1 的時候  L 也會有一個改變 

因為當 W1 改變 你就改變了 y, y 改變的時候你就改變了 e

什麼時候 W1 的改變會對 L 的影響很小? = 什麼時候 W1 這邊的變化 它在 error surface 上的斜率會很小?

ANS:當你的 input 很小的時候

假設 x1 的值在不同的 training example 裡面它的值都很小

因為 x1 是直接乘上 w1, 如果 x1 的值都很小, W1 有一個變化的時候它對 y 的影響也是小的
它對 y 的影響是小的 -> 對 e 的影響也是小的->它對 L 的影響就會是小的

假設 x2 的值都很大 當你的 W2 有一個小小的變化的時候 雖然 W2 這個變化可能很小
但是因為它乘上了 x2, x2 的值很大, 那 y 的變化就很, 那 e 的變化就很大->它對 L 的影響就會是大的

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/a1589985-43e2-4836-b41b-7ebb3a424546)

當我們 input 的 feature 它的 scale 差距很大的時候 我們就可能產生像這樣子的 error surface
就可能產生不同方向斜率非常不同的 error surface

那怎麼讓不同的 dimension有類似的 有接近的數值的範圍呢 => Normalization
![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/f62b0fb6-49d1-464e-a1a7-92e408b8de94)

共有R個Feature, 將同一個維度的資料進行運算得到平均, 之後再做standardization

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/2bda7f70-862e-4ce7-b782-391d982db6c6)

但是通過 W1 以後它就沒有做 normalize

如果 x tilde 通過 W1 得到是 z1
而 z1 不同的 dimension 間 它的數值的分布仍然有很大的差異的話
那我們要 train W2 第二層的參數 會不會也有困難呢
所以這樣想起來 我們也應該要對這邊的 a 或對這邊的 z做 Feature Normalization 
對 W2 來說 這邊的 a 或這邊的 z 其實也是一種 feature

Q:應該要在 activation function 之前做 normalization 還是要在 activation function 之後 normalization 呢
A:在實作上這兩件事情其實差異不大, 但如果用Sigmoid推薦對z做

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/64892087-fc66-4703-b2cb-5f363455b80a)

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/c0da849f-0b12-474e-947d-89a39f96704c)
當你改變 z1 的值的時候 μ 跟 Σ 也會跟著改變, 會造成後面跟著改變

之前的 network 都只吃一個 input 得到一個 output 
現在有一個比較大的 network 它是吃一堆 input
在這個 network 裡面 要算出 μ 跟 Σ 然後接下來產生一堆 output

# Batch Normalization 
在實作的時候 你不會讓這一個 network 考慮整個 training data 裡面的所有 example
你只會考慮一個 batch 裡面的 example 

舉例來說 你 batch 設 64
那你這個巨大的 network 就是把 64 筆 data 讀進去
算這 64 筆 data 的 μ和算這 64 筆 data 的 σ 對這 64 筆 data 都去做 normalization

因為我們在實作的時候 我們只對一個 batch 裡面的 data 做 normalization
所以這招叫做 Batch Normalization 

一定要有一個夠大的 batch才算得出 μ 跟 σ 
假設 batch size 設 1 那你就沒有什麼 μ 或 σ 可以算
所以這個 Batch Normalization 是適用於 batch size 比較大的時候

β 跟 γ 

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/a1eb2556-8501-4cfa-a2db-af63310a8410)

β 跟 γ 你要把它想成是 network 的參數 它是另外再被認出來的

Q:為什麼要加上 β 跟 γ 呢 
A:因為做 normalization 以後, 這邊的 z tilde 它的平均呢 就一定是 0 
那也許今天如果平均是 0 的話 就是給那 network 一些限制
那也許這個限制會帶來什麼負面的影響所以我們把 β 跟 γ 加回去
讓 network 現在它的 hidden layer 的 output 不需要平均是 0 
如果它想要平均不是 0 的話 他就自己去認這個 β 跟 γ
來調整一下輸出的分布 來調整這個 z hat 的分布 

Q:做 Batch Normalization 就是為了要讓每一個不同的 dimension它的 range 都是一樣
現在如果加去乘上 γ再加上 β 這樣不會不同 dimension 的分布它的 range 又都不一樣了嗎

A:有可能
但是實際上在做的時候啊   
γ 的初始值 就都設為 =1 γ 是一個 one vector都是 1 的向量
β 全部都是 0 的向量  = 一個 zero vector 裡面的值都是 0 的向量
所以讓你的 network 在一開始訓練的時候每一個 dimension 的分布是比較接近的

也許訓練到後來 你已經訓練夠長的一段時間已經找到一個比較好的 error surface

走到一個比較好的地方以後 那再把 γ 跟 β 慢慢地加進去
所以加 Batch Normalization 往往對你的訓練是有幫助的

# Batch normalization – Testing 

testing = inference 

在 testing 的時候你確實也可以在 testing 的資料上面製造一個一個 batch 

但是假設是一個真正的線上的 application, 一筆資料進來就要可以做預測

μ 跟 Σ 是用一個 batch 的資料算出來的

但如果今天在 testing 的時候 根本就沒有 batch 那我們要怎麼算這個 μ Σ 呢?

實作上的解法:moving average(PyTorch 幫你處理好了)

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/eae2330e-bd16-4d84-9b4c-a18790a31be8)

如果你有在做 Batch Normalization 的話 
PyTorch 在 training 的時候 你每一個 batch 計算出來的 μ 跟 Σ他都會拿出來算 moving average 

你每一次取一個 batch 出來的時候你就會算一個 μ1 
取第二個 batch 出來的時候 你就算個 μ2 
一直到取第 t 個 batch 出來的時候你就算一個 μ t 

接下來你會算一個 moving average 算出來 μ 的一個平均值

![image](https://github.com/Pin-Jiun/Machine-Learing-NTU/assets/34474924/64ac3ce9-2085-4d3b-a6ea-aefc7fb7e985)

相較於沒有做 Batch Normalization 只需要一半 或甚至更少的時間就跑到同樣的正確率了
